runtime/src/commonMain/kotlin/pbandk/internal/TimeUtil.kt

```diff
@@ -358,3 +359,53 @@ internal fun parseTime(str: String): Timestamp {
 }
 
 // endregion
+
+// region duration
+
+internal fun formatDuration(dur: Duration): String {
+    var seconds = dur.seconds
+    var nanos = dur.nanos
+
+    return buildString {
+        if (seconds < 0 || nanos < 0) {
+            append('-')
+            seconds = -seconds
+            nanos = -nanos
+        }
+
+        require(seconds <= 315576000000) { "Duration exceeds limits of +/- 315576000000 seconds" }
+
+        append(seconds)
+        if (nanos != 0) append(".${formatNanos(nanos)}")
+        append('s')
+    }
+}
+
+internal fun parseDuration(str: String): Duration {
+    require(str.isNotEmpty() && str.last() == 's') { "Expected duration to end with s" }
+
+    val negative = str.first() == '-'
+    val signLength = if (negative) 1 else 0
+
+    var seconds: Long
+    var nanos: Int
+    val nanosPos = ParsePosition(str.lastIndexOf('.') + 1)
+    if (nanosPos.position == 0) {
+        seconds = str.substring(signLength, str.length - 1).toLong()
+        nanos = 0
+    } else {
+        seconds = str.substring(signLength, nanosPos.position - 1).toLong()
+        nanos = parseNanos(str, nanosPos)
```

Comment:
🔧 You should `require(nanosPos.position == str.length - 1)` here, to catch the case of extra garbage characters between the nanos and the `'s'` at the end.